src/main/java/io/vertx/core/net/impl/KeyStoreHelper.java

```diff
@@ -112,34 +118,118 @@ public static KeyStoreHelper create(VertxInternal vertx, TrustOptions options) {
 
-  public KeyStoreHelper(String password) {
+  public KeyStoreHelper(KeyStore ks, String password) throws Exception {
+    Enumeration<String> en = ks.aliases();
+    while (en.hasMoreElements()) {
+      String alias = en.nextElement();
+      Certificate cert = ks.getCertificate(alias);
+      if (cert instanceof X509Certificate) {
+        X509Certificate x509Cert = (X509Certificate) cert;
+        Collection<List<?>> ans = x509Cert.getSubjectAlternativeNames();
+        List<String> domains = new ArrayList<>();
+        if (ans != null) {
+          for (List<?> l : ans) {
+            if (l.size() == 2 && l.get(0) instanceof Number && ((Number) l.get(0)).intValue() == 2) {
+              String dns = l.get(1).toString();
+              domains.add(dns);
+            }
+          }
+        }
+        String dn = x509Cert.getSubjectX500Principal().getName();
+        LdapName ldapDN = new LdapName(dn);
+        for (Rdn rdn : ldapDN.getRdns()) {
+          if (rdn.getType().equalsIgnoreCase("cn")) {
+            String name = rdn.getValue().toString();
+            domains.add(name);
+          }
+        }
+        if (domains.size() > 0) {
+          PrivateKey key = (PrivateKey) ks.getKey(alias, password != null ? password.toCharArray() : null);
+          Certificate[] tmp = ks.getCertificateChain(alias);
+          if (tmp == null) {
+            // It's a private key
+            continue;
+          }
+          List<X509Certificate> chain = Arrays.asList(tmp)
+              .stream()
+              .map(c -> (X509Certificate)c)
+              .collect(Collectors.toList());
+          X509KeyManager mgr = new X509KeyManager() {
+            @Override
+            public String[] getClientAliases(String s, Principal[] principals) {
+              throw new UnsupportedOperationException();
+            }
+            @Override
+            public String chooseClientAlias(String[] strings, Principal[] principals, Socket socket) {
+              throw new UnsupportedOperationException();
+            }
+            @Override
+            public String[] getServerAliases(String s, Principal[] principals) {
+              throw new UnsupportedOperationException();
+            }
+            @Override
+            public String chooseServerAlias(String s, Principal[] principals, Socket socket) {
+              throw new UnsupportedOperationException();
+            }
+            @Override
+            public X509Certificate[] getCertificateChain(String s) {
+              return chain.toArray(new X509Certificate[chain.size()]);
+            }
+            @Override
+            public PrivateKey getPrivateKey(String s) {
+              return key;
+            }
+          };
+          for (String domain : domains) {
+            if (domain.startsWith("*.")) {
+              wildcardMgrMap.put(domain.substring(2), mgr);
+            } else {
+              mgrMap.put(domain, mgr);
```

Comment:
Should we perhaps check the return values of the two `put()` calls here and check they are null to detect (and report) accidentally loading multiple certificates for the same domain(s)?